File: Javascript/Algorithms/1038.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
let bstToGst = root => {
    const dfs_inorder = (node, s) => {
        const {val, left, right} = node;
        
        // Go right and accumulate value in s.
        if (right){
            let [nr, ns] = dfs_inorder(right, s);
            s = ns
            node.right = nr;
        }
        node.val = val + s;
        // Need to recurse into left and return accumulated value.
        // Need to *still* return node though (returning node.left)
        // would prune almost all of the right subtree.
        if (left){
            let [nl, ns] = dfs_inorder(left, node.val);
            node.left = nl;
            return [node, ns];
        }
        return [node, node.val];
    }

    if (root)
        return dfs_inorder(root, 0)[0];
    return root;
};

File: C/Algorithms/1038.c
//  Definition for a binary tree node.
struct TreeNode {
  int val;
  struct TreeNode *left;
  struct TreeNode *right;
};
typedef struct TreeNode Node;

void dfs_inorder(Node *node, int *s) {
  Node *left = node->left, *right = node->right;

  // Recurse right, alters right->val and sets s
  // pointer to new value.
  if (right) {
    dfs_inorder(right, s);
  }
  // Update node value.
  node->val += *s;
  if (left) {
    // Recurse with s set to new node-val and then set s
    // to whatever val has been altered to.
    int val = node->val;
    dfs_inorder(left, &val);
    *s = val;
  } else {
    // Else, just set it to node->val
    *s = node->val;
  }
}

struct TreeNode *bstToGst(struct TreeNode *root) {
  if (root) {
    int s = 0;
    dfs_inorder(root, &s);
  }
  return root;
}
